Cython/Compiler/Nodes.py

```diff
@@ -2184,6 +2184,8 @@ def assure_gil(code_path, code=code):
                 # code.put_trace_exception()
 
                 assure_gil('error')
+                if code.funcstate.error_without_exception:
+                    code.put("if (PyErr_Occurred()) ")
                 code.put_add_traceback(self.entry.qualified_name)
```

Comment:
This is making the non-local assumption that `put_add_traceback()` issues a single command. If that ever stops being true, this code may lead to subtle bugs. Curly braces are cheap in C, let's just use them.

Also, we may be able to use `__Pyx_PyErr_Occurred()` here, which avoids a lookup of the thread state (if it has been looked up before, so not sure if this always applies).

An alternative might be a C flag variable that remembers when no exception was set.

Reply:
I considered C flag variables and decided it was a little more complicated, although probably more efficient. There's a third option that might work though:

currently the code generated is:
```
error_label:
Py_XDECREF(cleanup_code)
__Pyx_Add_Traceback(....)
```

If we could reorder it to:

```
error_label:
__Pyx_AddTraceback(...)
stop_iteration_label:
__Pyx_XDECREF(variable_cleanup)
```

then that'd be very simple. I wasn't sure if there was a good reason to add the traceback after the cleanup though.

Reply:
I couldn't immediately see how to use `__Pyx_PyErr_Occurred` (I think I need to generate a threadstate temp?) so I've left this suggestion for the moment.

But added the curly braces

Reply:
I've switched to using a C flag variable now.

On a quick microbenchmark (essentially a simplified version of `list(reimplementation_of_range(10))` I can't see any difference between a flag variable and `PyErr_Occurred()`. Both of those give a 30% speed up vs the current Cython master and a 90% speedup vs plain Python.

Therefore the change is worthwhile I think, but it isn't worth worrying about exactly how we detect the lack of error.